[STEP 1] redmango, dasanKim

DasanKim opened the pull request:
@seunghyunCheon 

# [STEP 1] redmango, dasanKim

## 🥭 고민되었던 점
### 짝프로그래밍
> 생소하고 사전협의 없이 바로 시작해야하니까 머리도 새하얘지고 했죠 by redmango
- 짝프로그래밍이 처음이라 처음엔 10~15분씩 돌아가면서 함 => commit 단위가 애매해짐
- 따라서 기능별로 돌아가면서 짝프로그래밍을 함
- 로웬과 브로디의 조언으로 시간을 늘려가면서 commit 단위를 맞춰보려고 합니다!

### 중복되지 않은 값 Collection Type - Set vs. Array
1. 중복되지 않은 값을 넣기 위해 `Set`을 쓸까
    - Set: 중복되지 않은 멤버가 순서없이 존재하는 컬렉션
    - 따라서 print할 경우 삽입 순서에 상관없이 뒤죽박죽 출력됨.
2. `Array`를 써서 중복된 값을 뺄까
    - Array: 멤버가 순서(인덱스)를 가진 컬렉션

=> 중복 체크를 축약하기 위해서 Set으로 선택 후 순서를 위해 Array로 데이터 형변환

### 네이밍 고민
네이밍은 언제나 고민되는 일이다
1. remainingChance / remainingCount
    - 보통 count를 쓰지만, 야구게임인 것을 감안하여 chance로 선택함
2. compare 함수의 전달인자 이름을 and, to, with 중에 어떤 전치사를 사용할지 고민함.
    - compare(A ,and B)
    - A와 B가 동등한 상태에서 비교된다는 점을 감안하여 and 선택함.
3. computerRandomNumbers -> hiddenRandomNumbers
   - computer라는 단어가 적절한 것 같지 않아, 숨겨진 값이라는 점을 부각하기 위하여 hiddenRandomNumbers으로 정함.

### 전역변수 사용 고민
기존에 전역변수로 선언했던 
- hiddenRandomNumbers 
- userRandomNumbers
- remainingChance

위 3가지 변수들의 사용처가 제한적이여서 전역변수로 사용할 필요가 없다고 판단하여, 지역변수로 변환함.

## 🍓 조언을 얻고 싶은 부분
### 고차함수 활용
``` Swift
func compare(_ hiddenRandomNumbers: [Int], and userNumbers: [Int]) -> (strikeCount: Int, ballCount: Int) {
    var strikeCount: Int = 0
    var ballCount: Int = 0
    
    for index in 0...2 {
        if hiddenRandomNumbers[index] == userNumbers[index] {
            strikeCount += 1
        } else if hiddenRandomNumbers.contains(userNumbers[index]){
            ballCount += 1
        }
    }
    return (strikeCount: strikeCount, ballCount: ballCount)
}
```
에서 for문 내부 부분을 enumerated()과 filter()를 사용하여 고차함수로 구현할 수 있지 않을까 고민했으나
현재 코드가 가독성과 단순한 면에서 더 좋다고 판단하여 유지하였습니다.
저희의 판단이 옳았을까요?🤔

DasanKim committed:
feat: 전역변수 및 랜덤한정수생성하는 함수 추가

Commit:
feat: 랜덤한 정수를 생성하는 함수 구현

DasanKim committed:
feat: 숫자비교함수 추가

Commit:
feat: 게임 시작 함수 일부 추가

DasanKim committed:
refactor: 네이밍 및 startGame() 함수 수정

Commit:
refactor: 함수 및 변수명 변경, 일부 코드 수정

전역변수의 필요성이 없기에 지역변수로 변경,

랜덤한 정수를 생성하는 함수의 기존 중복체크 코드를 축약하기 위해 Set을 사용하여 변경

임의의 수 출력시 고차함수 활용을 위해 Map으로 변경

DasanKim committed:
docs: README.md에 순서도 추가

seunghyunCheon commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,10 +1,60 @@
+}
+
+func startBaseballGame() {
+    let hiddenRandomNumbers = creatRandomNumbers()
```
네이밍을 고민한 점은 좋습니다! 👍
다만 숨겨진 값을 부각하기 위한 값이라는 PR내용을 보기 전에 파일을 열어보았을 때는 의미가 조금 모호하다고 생각이 들었어요. `computerRandomNumbers`가 적절하지 않다고 생각한 이유를 들어볼 수 있을까요?

DasanKim replied:
`computerRandomNumbers`가 적절하지 않다고 생각하기보다 “유저가 숨겨진 번호를 맞춘다는 행동”에 초점을 둬서 그런 것 같습니다:) 브로디가 질문을 해주신 이후에 함께 고민을 해보았는데요, “유저의 행동”이라는 1인칭 시점보다 “게임을 누구와 하느냐”로 3인칭 시점에서 생각해보았을 때 유저와 컴퓨터의 대결이므로 `computerRandomNumbers`가 `hiddenRandomNumbers`보다 직관적인 것 같다는 결론을 내렸습니다. 생각할 수 있는 질문을 주셔서 감사해요 브로디!😁

seunghyunCheon commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,10 +1,60 @@
 import Foundation
 
-print("Hello, World!")
+func creatRandomNumbers() -> [Int] {
```
`creat`는 오타가 맞겠죠? 🙃 
`Xcode -> Edit -> Format -> spelling and Grammar`에서 스펠링을 체크하는 옵션이 있답니다!

DasanKim replied:
알려주신대로 스펠링 체크하는 옵션을 실행해봤더니 바로 creat를 잡아주네요! 
감사합니다😆

seunghyunCheon commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,10 +1,60 @@
-// 
+//
 
 import Foundation
```
`Foundation`은 어떤 역할을 하나요?
이 파일에서 꼭 필요한 것 일까요?

DasanKim replied:
Foundation의 역할과 필요성 
 1. 시간 및 날짜 계산 
 2. 텍스트 가공 
 3. 정렬 및 필터링 
 4. 네트워킹 
 5. 데이터 저장소 등의 기능을 제공한다. 

지금 이 코드에선 `Swift Standard Library의 기능으로 충분`하기 때문에 사용할 필요가 없다고 판단하였습니다!

seunghyunCheon commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,10 +1,60 @@
+        
+        let gameResult = compare(hiddenRandomNumbers, and: userRandomNumbers)
+        let strikeCount: Int = gameResult.strikeCount
+        let ballCount: Int = gameResult.ballCount
```
타입 어노테이션과 타입 추론을 혼용해서 사용하고 계신데 기준이 있으실까요?
컨벤션을 통일 시켜주면 좋을 것 같은 생각입니다 😀

DasanKim replied:
기준이 없어서 이번에 만들었습니다:)
타입 추론 대신 `타입 어노테이션을 사용`하기로 컨벤션을 통일하였습니다☺️
감사합니다!

seunghyunCheon commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,10 +1,60 @@
+        } else if hiddenRandomNumbers.contains(userNumbers[index]){
+            ballCount += 1
+        }
+    }
```
`compare`함수는 숫자가 3개인 경우에만 정상적으로 동작하기에 만약 요구사항이 변한다면(ex. 6개의 값을 검사해야 한다면) 특정 부분을 바꿔야겠네요. 
특정 개수에 의존하지않는 방법을 생각해 볼까요?

seunghyunCheon commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,10 +1,60 @@
+        } else {
+            print("남은 기회: \(remainingChance)")
+        }
+    }
```
`startBaseballGame`함수 안에서 게임 전체흐름을 한 눈에 볼 수 있게끔 승리판별 기능을 함수로 분리해도 좋을 것 같아요.

seunghyunCheon commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,10 +1,60 @@
 
-print("Hello, World!")
+func creatRandomNumbers() -> [Int] {
+    var uniqueRandomNumbers = Set<Int>()
```
Set에 적합한 네이밍 이네요👍

seunghyunCheon commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,10 +1,60 @@
+    return Array(uniqueRandomNumbers)
+}
+
+func compare(_ hiddenRandomNumbers: [Int], and userNumbers: [Int]) -> (strikeCount: Int, ballCount: Int) {
```
`argumentLabel`을 잘 사용하셨네요 👍

seunghyunCheon commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,10 +1,60 @@
+        let strikeCount: Int = gameResult.strikeCount
+        let ballCount: Int = gameResult.ballCount
+        
+        print("임의의 수 : \(userRandomNumbers.map{String($0)}.joined(separator: " "))")
```
```swift
print("임의의 수 : \(userRandomNumbers.map{ String($0) }.joined(separator: " "))")
```
일반적으로 고차함수는 중괄호 안을 띄어씁니다.

seunghyunCheon commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,10 +1,60 @@
+}
+
+func startBaseballGame() {
+    let hiddenRandomNumbers = creatRandomNumbers()
+    var userRandomNumbers = [Int]()
+    var remainingChance: Int = 9
```
해당 변수들을 지역변수로 사용해주셨네요! 
지역변수와 전역변수의 장단점은 무엇일까요? (시간이 되신다면 메모리 측면에서도 살펴보시면 좋을 것 같아요!)

DasanKim replied:
## 지역변수(local variable)   
- 특정구역({}) 내에서 생성되는 변수로 보통 함수 내에 선언되어 해당 함수 안에서만 사용가능하다.   
- Stack Area에 저장되어 함수(코드블록)가 종료되면 저장된 메모리도 자동으로 해제된다.   
- 즉, 함수가 종료되면 함수 내에 있던 전역변수들의 값은 사라진다. 

## 전역변수(Global variable)   
- 함수 밖에 선언하여 모든 지역에서 사용가능한 변수   
- Data Area에 저장되어 프로그램이 시작할 때 할당되고, 프로그램이 종료되어야 메모리에서 해제된다.   
### 장점      
- 이름처럼 전역적으로 사용할 수 있고,       
- 지역변수와 다르게 프로그램이 종료될 때까지 저장된 값이 사라지지 않는다.   
### 단점: 장점이 단점이 될 수 있다.       
- 전역적으로 사용되어 결과를 예측하기 어렵고, 프로그램의 규모가 커질수록 위험성이 높아지며 추적이 힘들 수 있다.      
- 프로그램이 종료될 때까지 메모리 공간을 차지하기 때문에, 메모리 부족의 원인이 될 수 있다. 

=> 전역변수는 프로그램의 구조를 복잡하게 만들고 사용빈도와 상관없이 프로그램이 끝날 때까지 메모리를 차지하고 있기 때문에 전역변수를 사용하는 것은 신중해야함. 

요렇게 정리해보았습니다:)
[참고] https://edu.goorm.io/learn/lecture/2010/%ED%95%9C-%EB%88%88%EC%97%90-%EB%81%9D%EB%82%B4%EB%8A%94-%EA%B3%A0%EB%9E%AD-%EA%B8%B0%EC%B4%88/lesson/174452/%EC%A0%84%EC%97%AD%EB%B3%80%EC%88%98%EC%99%80-%EC%A7%80%EC%97%AD%EB%B3%80%EC%88%98

seunghyunCheon reviewed (commented):
안녕하세요! redmango, dasanKim 🙌
첫 번째 PR작성하시느라 고생하셨습니다 😊
전체적으로 코드가 보기편했고 기능분리가 잘 되어있다는 생각이 들었습니다! 순서도도 정말 잘 작성해주셨네요 😄
코드에는 정답이 없기 때문에 제 의견이 정답도 아니고 필히 수정해야 하는 부분도 아닙니다. 
해당 코드를 작성한 명확한 이유나 근거가 있다면 저에게도 가르쳐주시면 좋겠습니다 😀

고민하신 점과 관련하여 코멘트로 남겨 드렸습니다.
아래는 질문에 대한 답변입니다.

### 고차함수 활용

판단의 옳고그름을 내릴 수는 없지만 제 생각은 여러분과 같습니다.
`enumerated`와 `filter`를 사용한다면 간결함은 증가하겠지만 가독성은 떨어질 것 같습니다. 그리고 현재 상태에서 해당 고차함수의 사용이 코드 수를 많이 줄일지에 대해서도 조금 의문이 드네요! 

지금의 상태가 좋아보입니다! 그리고 고차함수를 사용하신다면 축약문법을 적극적으로 사용하기보다는 현재 사용하고 있는 변수가 무엇인지, 이 변수의 네이밍을 무엇으로 할 지에 집중하는 것이 좋아보여요!

Commit:
refactor: 일부 수정사항 반영

creat 오타수정
고차함수 Map 컨벤션 적
startBaseballGame 함수 내 불필요한 변수 삭제
compare 함수내 반복문 반복횟수를 특정 갯수가 아닌 배열 크기에 의존하도록 수정

DasanKim committed:
refactor: 승부판정 기능 분리

startBaseballGame함수에서 승부판정 기능을 분리하여 decideWinner 함수 추가

DasanKim committed:
refactor: 함수명 및 변수명 변경

DasanKim committed:
refactor: 게임시작 함수 및 승부판정 함수 수정, 변수명 수정

DasanKim commented:
안녕하세요 브로디(@seunghyunCheon)!

브로디 덕분에 생산적인 고민을 할 수 있어서 감사했습니다!☺️

고민해봤던 부분들은 댓글로 남겼구요
수정한 사항들은 아래와 같습니다.

1. compare함수에서 특정 개수에 의존하지 않도록 수정
2. 승부판별 기능을 함수인 judgeWinner 함수 추가
3. 고차함수는 중괄호 안 띄어쓰기 반영
4. 그 외 변수명 수정

특별히 승부판별하는 `함수명과 그 함수의 전달인자 레이블명`을 정하는데 고민이 많았습니다^_ㅠ
게임시작 함수 내에서 선언적인 코드를 작성하여 직관성을 높여보려고 노력했습니다!

오늘도 감사합니다 브로디!🥭
